### PR TITLE
[FEQ]Yaswanth/FEQ-1707_ Added badge component

### PR DESCRIPTION
**Pull request by yaswanth-deriv**
I've enhanced the deriv-com/ui library by introducing the badge component, Added below props to use all variants
Color: We now offer a selection of six colors, addressing any previously missing options in the roos.css file.
Size: Users can choose from extra-small, small, medium, and large sizes to suit their layout requirements.
Variation: Two distinct styles, 'contained' and 'bordered', are available for users to apply as per their design preferences.
Padding: Flexible padding options of 'loose' and 'tight' ensure optimal spacing within the badge.
Font Boldness: Users have the flexibility to emphasize text within the badge by toggling bold font weight.
Roundness: The badge's edges can be customized with 'full', 'medium', or 'small' roundness options.
Icons: Optional right and left icons allow for further customization and visual enhancement.
Children: Users can seamlessly incorporate additional elements or text within the badge using ReactNode.

Screenshot 2024-03-15 at 11 34 34 AM
Screenshot 2024-03-15 at 11 34 58 AM

### CODE REVIEW

File: src/components/Badge/Badge.scss [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    justify-content: center;
+    align-items: center;
+    position: relative;
+
+    &__size {
+        &--tight {
+            &.deriv-badge__padding--xs {
+                width: 34px;
+                height: 14px;
+                gap:4px;
```
Comment (nada-deriv) [inline src/components/Badge/Badge.scss:31]: `gap: 4px;`

File: src/components/Badge/__test__/Badge.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import { Badge } from '..';
+
+describe('Badge component', () => {
+  it('renders children', () => {
+    render(<Badge isBold={false}>Test</Badge>);
+    expect(screen.getByText('Test')).toBeInTheDocument();
+  });
+
+  it('renders left and right icons', () => {
+    const leftIcon = <span data-testid="left-icon">Left Icon</span>;
```
Comment (nada-deriv) [inline src/components/Badge/__test__/Badge.spec.tsx:12]: `data-testid='dt_.....'`

**Comment by ameerul-deriv** [comment]
I think there are some slight design differences with P2P cause the colours are slightly different. Here are the different badges we have and the different colour codes

### Gold
Screenshot 2024-03-18 at 10 13 51 AM

### Success
Screenshot 2024-03-18 at 10 14 01 AM

### General
Screenshot 2024-03-18 at 10 14 08 AM

### Green
Screenshot 2024-03-18 at 10 16 31 AM

### Colour Codes
Screenshot 2024-03-18 at 10 16 51 AM

**Comment by yaswanth-deriv** [comment]
Sure, I've matched the color codes from the Figma file. I'll add these colors accordingly. Thanks!

File: src/components/Badge/index.tsx [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+} as const;
+
+const FontSize = {
+    lg: "lg",
+    md: "md",
+    sm: "sm",
+    xs: "xs",
+} as const;
+
+export const Badge = ({
```
Comment (aizad-deriv) [inline src/components/Badge/index.tsx:68]: can we add forwardref for this

File: src/components/Badge/Badge.scss [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+$success_color: $gradient-color-blue-5;
+$warning_color: $gradient-color-yellow;
+$danger_color: var(--du-badge-danger, #ec3f3f);
+$blue_color: var(--du-badge-blue, #115BFF);
+$purple_color: var(--du-badge-purple, #722FE4);
+$light_blue_color: var(--du-badge-light-blue, #0DC2E7);
+$gold_color:$gradient-color-gold;
+$gold_color_border:$color_yellow;
+$green_color:$gradient-color-green-4;
+$green_color_border:$color-green-6;
```
Comment (aizad-deriv) [inline src/components/Badge/Badge.scss:15]: Nothing wrong with this but can be utilize even more by using scss maps.

**Review by aizad-deriv — APPROVED** [review]

File: src/components/Badge/Badge.scss [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    display: inline-flex;
+    justify-content: center;
+    align-items: center;
+    position: relative;
+
+    &__size {
+        &--tight {
+            &.deriv-badge__padding--xs {
+                width: 34px;
+                height: 14px;
```
Comment (jim-deriv) [inline src/components/Badge/Badge.scss:27]: @yaswanth-deriv, do you guys (`@deriv/ui`) typically work with `px` as units for such? 🤔 
```suggestion
                width: 34px;
                height: 14px;
```

File: src/components/Badge/Badge.scss [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        }
+
+        &--bold-text {
+            text {
+                font-weight: bold;
+            }
+        }
+    }
+
+}
```
Comment (jim-deriv) [inline src/components/Badge/Badge.scss:251]: Could you please address the `EOF` @yaswanth-deriv?

File: src/components/Badge/__test__/Badge.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import { Badge } from "..";
+
+describe("Badge component", () => {
+    it("renders children", () => {
+        render(<Badge isBold={false}>Test</Badge>);
+        expect(screen.getByText("Test")).toBeInTheDocument();
+    });
+
+    it("renders left and right icons", () => {
+        const leftIcon = <span data-testid="left-icon">Left Icon</span>;
```
Comment (jim-deriv) [inline src/components/Badge/__test__/Badge.spec.tsx:12]: Do you guys not use the `dt` prefix @yaswanth-deriv?
```suggestion
        const leftIcon = <span data-testid="left-icon">Left Icon</span>;
```

File: src/components/Badge/index.tsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    | "purple"
+    | "success"
+    | "warning"
+    | "danger"
+    | "gold"
+    | "green"
+type TPadding = "tight" | "loose";
+
+interface BadgeProps extends ComponentProps<"div"> {
+    children?: ReactNode;
```
Comment (jim-deriv) [inline src/components/Badge/index.tsx:21]: `ReactNode` already includes `undefined` @yaswanth-deriv.
```suggestion
    children: ReactNode;
```

File: src/components/Badge/__test__/Badge.spec.tsx (hunk 7 above)
Comment (yaswanth-deriv) [reply src/components/Badge/__test__/Badge.spec.tsx:12]: We use dt but as it is test cases I used this format

File: src/components/Badge/Badge.scss (hunk 5 above)
Comment (yaswanth-deriv) [reply src/components/Badge/Badge.scss:27]: Yes, we are consistently using px throughout the library

File: src/components/Badge/index.tsx [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+interface BadgeProps extends ComponentProps<"div"> {
+    children: ReactNode;
+    color?: TColor;
+    isBold: boolean;
+    rightIcon?: ReactElement;
+    leftIcon?: ReactElement;
+    padding?: TPadding;
+    badgeSize?: Extract<TGenericSizes, "lg" | "md" | "sm" | "xs">;
+    textSize?: ComponentProps<typeof Text>["size"];
+    variant?: TVariant;
```
Comment (shayan-deriv) [inline src/components/Badge/index.tsx:29]: we have three different border-radius. so we need another prop for that

**Review by ameerul-deriv — APPROVED** [review]

File: src/components/Badge/Badge.scss
Comment (shayan-deriv) [inline src/components/Badge/Badge.scss:1]: I think you missed the mobile design ( or if you went with mobile-first approach, then u missed the desktop design ).

Comment (shayan-deriv) [reply src/components/Badge/Badge.scss:1]: image

Comment (yaswanth-deriv) [reply src/components/Badge/Badge.scss:1]: Thanks I have added it.Please check now

File: src/components/Badge/index.tsx [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    | "danger"
+    | "gold"
+    | "green"
+type TPadding = "tight" | "loose";
+
+interface BadgeProps extends ComponentProps<"div"> {
+    badgeSize?: Extract<TGenericSizes, "lg" | "md" | "sm" | "xs">;
+    children: ReactNode;
+    color?: TColor;
+    isBold: boolean;
```
Comment (shahzaib-deriv) [inline src/components/Badge/index.tsx:24]:
```suggestion
    isBold?: boolean;
```

Is it expected to make `isBold` a required prop?

Comment (yaswanth-deriv) [reply src/components/Badge/index.tsx]: I am making it optional.Thanks

File: src/components/Badge/Badge.scss [hunk 11]
```diff
@@ -0,0 +1,511 @@
...
+            border-radius: 8px;
+        }
+        &--lg {
+            border-radius: 24px;
+        }
+    }
+
+}
+
+@include mobile{
```
Comment (shayan-deriv) [inline src/components/Badge/Badge.scss:264]: @yaswanth-deriv, given that we follow a mobile-first approach, could you please move these styles to make it the default style? For the desktop styles, place them within the "@include desktop"

**Review by shayan-deriv — COMMENTED** [review]
just a minor comment.
the rest LGTM

**Review by shayan-deriv — APPROVED** [review]

**Pull request merged**